### ISSUE TITLE
channelInactive shouldn't call before channelActive

### ISSUE THREAD

**Issue by andreypfau** [netty/netty#11291]
### Expected behavior
prints only "Closing channel..."

### Actual behavior
prints "Closing channel..." and "Channel now inactive!"

### Steps to reproduce
https://gist.github.com/andreypfau/27f7a5616a4d5bbe19526d87928482a3

### Netty version
4.1.65

### JVM version (e.g. `java -version`)
Java(TM) SE Runtime Environment GraalVM EE 21.1.0 (build 16.0.1+9-jvmci-21.1-b05)

### OS version (e.g. `uname -a`)
Darwin Kernel Version 20.4.0: Thu Apr 22 21:46:47 PDT 2021; root:xnu-7195.101.2~1/RELEASE_X86_64 x86_64

**Comment by NiteshKant** [comment]
Just for context; a handler in a pipeline is not guaranteed to receive all events on the channel because handlers may be added after certain events have already been processed by the pipeline. The only event ordering that is guaranteed is the order of events that happen on the pipeline, eg: `channelActive()` should not be called after it is closed.

In this case, the channel was closed during registration and hence you never received `channelActive()`.

**Comment by andreypfau** [comment]
This sounds logical, but judging by the documentation, channelInactive should only be called if the channel **WAS** active

**Comment by NiteshKant** [comment]
> channelInactive should only be called if the channel WAS active

You are correct but there is a distinction between channel being active and a handler receiving `channelActive()` callback. Here is the code which invokes `channelInactive()` during close:

https://github.com/netty/netty/blob/4.1/transport/src/main/java/io/netty/channel/AbstractChannel.java#L765

You can see that it only sends `channelInactive()` if the channel was active. So in this case, channel was active but the `channelActive()` event was not sent. Now, in order to make sure we do not send `channelInactive()` without sending `channelActive()` we would have to maintain the state of whether we have sent a `channelActive()` event before. Since this state may change per handler as pipeline changes can happen at any time, this state has to be managed per handler. Netty has opted for the trade-off of simplicity and less state management over lack of precision which is what you are observing in this case.
